include/swift/Driver/ExperimentalDependencyDriverGraph.h

```diff
@@ -0,0 +1,338 @@
+  /// which file the name was found.) In such a case, it is necessary to move
+  /// the node to the proper collection.
+  void moveNodeToDifferentFile(DriverNode *n, Optional<std::string> newFile) {
+    eraseNodeFromMap(n);
+    n->setSwiftDeps(newFile);
+    addToMap(n);
+  }
+
+  /// Remove node from nodeMap, check invariants.
+  DriverNode *eraseNodeFromMap(DriverNode *n) {
+    DriverNode *v = nodeMap.findAndErase(
+        n->getSwiftDeps().getValueOr(std::string()), n->getKey());
+    assert(
+        n == v ||
+        mapCorruption("Node found from key must be same as node holding key."));
+    return n;
+  }
+
+  static StringRef getSwiftDeps(const Job *cmd) {
+    return cmd->getOutput().getAdditionalOutputForType(
+        file_types::TY_SwiftDeps);
+  }
+
+  const Job *getJob(Optional<std::string> swiftDeps) const {
+    assert(swiftDeps.hasValue() && "Don't call me for expats.");
+    auto iter = jobsBySwiftDeps.find(swiftDeps.getValue());
+    assert(iter != jobsBySwiftDeps.end() && "All jobs should be tracked.");
+    assert(getSwiftDeps(iter->second) == swiftDeps.getValue() &&
+           "jobsBySwiftDeps should be inverse of getSwiftDeps.");
+    return iter->second;
+  }
+
+public:
+  /// For templates such as DotFileEmitter.
+  using NodeType = DriverNode;
+
+  DriverGraph() = default;
+
+  DependencyGraphImpl::LoadResult loadFromPath(const Job *, StringRef);
+
+  /// For the dot file.
+  std::string getGraphID() const { return "driver"; }
+
+  void forEachUseOf(const DriverNode *def,
+                    llvm::function_ref<void(const DriverNode *)>);
+
+  void forEachNode(llvm::function_ref<void(const DriverNode *)>) const;
+
+  void forEachArc(
+      llvm::function_ref<void(const DriverNode *, const DriverNode *)>) const;
+
+  /// Call \p fn for each node whose key matches \p key.
+  void forEachMatchingNode(const DependencyKey &key,
+                           function_ref<void(const DriverNode *)>) const;
+
+public:
+  // This section contains the interface to the status quo code in the driver.
+
+  /// Interface to status quo code in the driver.
+  bool isMarked(const Job *) const;
+
+  /// Visit closure of every use of \p job, adding each to visited.
+  /// Record any "cascading" nodes visited.
+  /// "Cascading" means has a use by an interface in another file.
+  void
+  markTransitive(SmallVectorImpl<const Job *> &visited, const Job *node,
+                 DependencyGraph<const Job *>::MarkTracer *tracer = nullptr);
+
+  /// "Mark" this node only.
+  bool markIntransitive(const Job *);
+
+  /// Record a new (to this graph) Job.
+  void addIndependentNode(const Job *);
+
+  std::vector<std::string> getExternalDependencies() const;
+
+  void markExternal(SmallVectorImpl<const Job *> &uses,
+                    StringRef externalDependency);
+
+  void verify() const;
+
+private:
+  bool static mapCorruption(const char *msg) {
+    llvm_unreachable(msg);
+    return false;
```

Comment:
Double nitpick: `static bool` vs. `bool static` and "no need for `return` after `llvm_unreachable`".

Reply:
Thanks, fixed.